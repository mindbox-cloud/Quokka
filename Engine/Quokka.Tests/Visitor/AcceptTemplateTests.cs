using System.Collections.Generic;

using Microsoft.VisualStudio.TestTools.UnitTesting;

using Snapshooter.MSTest;

namespace Mindbox.Quokka.Tests;

[TestClass]
public class AcceptTemplateTests
{
    [TestMethod]
    public void EnsureAllVisitMethodsAreInvokedForTemplate()
    {
		var templateText = """
							${ Math.Min(32, 16) + Math.Square(6) + Math.DecimalPart() }
							${ (24 + 3) + (5 * 25)/7 - (6 * 7 / 8*9) * (242) + A.Value * B.Value.Length }
							@{ set x = 353 + 255 }${ x }
							@{ if IsRed }
								Red
							@{ else if IsGreen }
								Green
							@{ else if IsBlue }
								Blue
							@{ else if IsYellow }
								Yellow
							@{ end if }
							@{ for item in Collection }
								List element
							@{ end for }
							""";

		var visitor = new TestTreeVisitor();
        var template = new Template(templateText);
		

        template.Accept(visitor);
		
		
        Snapshot.Match(visitor.VisitedNodes);
    }
}