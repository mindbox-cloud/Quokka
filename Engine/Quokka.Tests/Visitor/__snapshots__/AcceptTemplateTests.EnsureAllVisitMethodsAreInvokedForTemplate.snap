[
  "VisitTemplateBlock",
  "VisitStaticBlock",
  "VisitOutputInstructionBlock",
  "VisitAdditionExpression",
  "VisitPlusOperand",
  "VisitVariantValueArithmeticExpression",
  "VisitMemberValueExpression",
  "VisitVariableValueExpression: Math",
  "EndVisit",
  "VisitMethodMember: Min",
  "VisitArgumentValue",
  "VisitNumberExpression: 32",
  "EndVisit",
  "EndVisit",
  "VisitArgumentValue",
  "VisitNumberExpression: 16",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitPlusOperand",
  "VisitVariantValueArithmeticExpression",
  "VisitMemberValueExpression",
  "VisitVariableValueExpression: Math",
  "EndVisit",
  "VisitMethodMember: Square",
  "VisitArgumentValue",
  "VisitNumberExpression: 6",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitPlusOperand",
  "VisitVariantValueArithmeticExpression",
  "VisitMemberValueExpression",
  "VisitVariableValueExpression: Math",
  "EndVisit",
  "VisitMethodMember: DecimalPart",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitConstantBlock: \n",
  "EndVisit",
  "VisitOutputInstructionBlock",
  "VisitAdditionExpression",
  "VisitPlusOperand",
  "VisitAdditionExpression",
  "VisitPlusOperand",
  "VisitNumberExpression: 24",
  "EndVisit",
  "EndVisit",
  "VisitPlusOperand",
  "VisitNumberExpression: 3",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitPlusOperand",
  "VisitMultiplicationExpression",
  "VisitMultOperand",
  "VisitMultiplicationExpression",
  "VisitMultOperand",
  "VisitNumberExpression: 5",
  "EndVisit",
  "EndVisit",
  "VisitMultOperand",
  "VisitNumberExpression: 25",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitDivOperand",
  "VisitNumberExpression: 7",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitMinusOperand",
  "VisitMultiplicationExpression",
  "VisitMultOperand",
  "VisitMultiplicationExpression",
  "VisitMultOperand",
  "VisitNumberExpression: 6",
  "EndVisit",
  "EndVisit",
  "VisitMultOperand",
  "VisitNumberExpression: 7",
  "EndVisit",
  "EndVisit",
  "VisitDivOperand",
  "VisitNumberExpression: 8",
  "EndVisit",
  "EndVisit",
  "VisitMultOperand",
  "VisitNumberExpression: 9",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitMultOperand",
  "VisitNumberExpression: 242",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitPlusOperand",
  "VisitMultiplicationExpression",
  "VisitMultOperand",
  "VisitVariantValueArithmeticExpression",
  "VisitMemberValueExpression",
  "VisitVariableValueExpression: A",
  "EndVisit",
  "VisitFieldMember: Value",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitMultOperand",
  "VisitVariantValueArithmeticExpression",
  "VisitMemberValueExpression",
  "VisitVariableValueExpression: B",
  "EndVisit",
  "VisitFieldMember: Value",
  "EndVisit",
  "VisitFieldMember: Length",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitConstantBlock: \n",
  "EndVisit",
  "EndVisit",
  "VisitAssignmentBlock: x",
  "VisitAdditionExpression",
  "VisitPlusOperand",
  "VisitNumberExpression: 353",
  "EndVisit",
  "EndVisit",
  "VisitPlusOperand",
  "VisitNumberExpression: 255",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitStaticBlock",
  "VisitOutputInstructionBlock",
  "VisitVariableValueExpression: x",
  "EndVisit",
  "EndVisit",
  "VisitConstantBlock: \n",
  "EndVisit",
  "EndVisit",
  "VisitIfBlock",
  "VisitConditionBlock",
  "VisitTemplateBlock",
  "VisitStaticBlock",
  "VisitConstantBlock: \n\tRed\n",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitVariantValueBooleanExpression",
  "VisitVariableValueExpression: IsRed",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitConditionBlock",
  "VisitTemplateBlock",
  "VisitStaticBlock",
  "VisitConstantBlock: \n\tGreen\n",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitVariantValueBooleanExpression",
  "VisitVariableValueExpression: IsGreen",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitConditionBlock",
  "VisitTemplateBlock",
  "VisitStaticBlock",
  "VisitConstantBlock: \n\tBlue\n",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitVariantValueBooleanExpression",
  "VisitVariableValueExpression: IsBlue",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitConditionBlock",
  "VisitTemplateBlock",
  "VisitStaticBlock",
  "VisitConstantBlock: \n\tYellow\n",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitVariantValueBooleanExpression",
  "VisitVariableValueExpression: IsYellow",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitStaticBlock",
  "VisitConstantBlock: \n",
  "EndVisit",
  "EndVisit",
  "VisitForBlock: item",
  "VisitTemplateBlock",
  "VisitStaticBlock",
  "VisitConstantBlock: \n\tList element\n",
  "EndVisit",
  "EndVisit",
  "EndVisit",
  "VisitVariableValueExpression: Collection",
  "EndVisit",
  "EndVisit",
  "EndVisit"
]
